### PR TITLE
Off-chain group construction and management

### PR DESCRIPTION
**Pull request by staheri14**
This PR adds support for a static group formation in waku rln-relay and closes https://github.com/status-im/nim-waku/issues/713.
The idea is that a list of 50 membership keys in the form of (identity key, identity commitment key) is hardcoded in the waku node. When mounting rln-relay, the Merkle tree gets constructed out of this list. The root of the Merkle tree will be used to verify the non-spamming proof of the relayed messages. 
A node can pick up one of these hardcoded keys through the CLI option of `rln-relay-membership-index`. The selected keys are going to be used to generate proof for that node. 

Please note that the static group feature is not yet fully functional, that is we cannot start dogfooding as the proof generation and verification are not yet integrated. The proof parts will be incorporated in the next PR.

To test this feature, first compile wakunode2 with RLN enabled:
```
make wakunode2 RLN=true 
```
Then run the compiled code with the following CLI options: (you may change `rln-relay-membership-index` to other values between 0-49)
```
./build/wakunode2  --rln-relay=true --rln-relay-membership-index=1
```
The selected identity key and the identity commitment key are displayed under the following info logs. If you set `--rln-relay-membership-index` to 1, then your id key and id commitment key should be exactly like below:

```bash
DBG 2021-09-16 14:57:03.797-07:00 rln-relay membership index                 topics="wakunode" tid=876620 file=wakunode2.nim:595 **rlnRelayMemIndex=1**
INF 2021-09-16 14:57:07.507-07:00 membership id key                          topics="wakunode" tid=876620 file=wakunode2.nim:609 **idkey=27b2bfc25257e53819beaf36ce1070007e04e7aad2e440a1f1fc066f59a61123**
INF 2021-09-16 14:57:07.507-07:00 membership id commitment key               topics="wakunode" tid=876620 file=wakunode2.nim:610 **idCommitmentkey=522ce51aff96041e79a8476f508fb9661f146f189e288f83cb4837517cfc0127**
```

You can also see the Merkle tree root in the following log. The root value also should be the same as below:
```bash
DBG 2021-09-16 14:57:07.507-07:00 the calculated root                        topics="wakunode" tid=876620 file=wakunode2.nim:620 root=25e03b7d551c3016cb21c057a113be5bf27eaa9b2529e806715bc0a1ef221e07
```

### CODE REVIEW

File: tests/v2/test_waku_rln_relay.nim [hunk 1]
```diff
@@ -259,6 +285,37 @@ procSuite "Waku rln relay":
...
...1a8ae18bad00c35597010a2107cad5041ffab69329", "d73e8c0e36e5588836bc9be4820e108d008f8399e4a9bacb5a452d4850a72d10", "01b81a331bc0c7818812c3848d382551183b35a16d74df5c8c32d561edb74b1c", "403e3cb57be08aef37dcc8af0a93a6ea8064540c8bcdb76fb34875c3d894be03", "6e420bc668f7b2aeee7e719daa3a7a4471bd04e8bcf24a27e4a5ae07e8c44207", "98baf438da4c655b9d92dc9033ae0f5cd7f0efda9b0d1bb894d08f1ce8620a01", "1282dda3446fb157e236155197def0d309eecb9dde8445f599ead69d4f31d916", "6453a3cf73ab5f3bafe645ce69fcdb612526fa66659e564c783c5600fb3a400f", "a8c1bfdbae2f56d10ccaf89887286c7c3256f7bdba21e98ac3c44578926c642c", "34bd74a58df94ae160f90bb64d202908d95b9dd0d89f02cd7e6bc8ddaf11280f", "a7377899f99a93796cd8901e17357403a900570d869fc49afed5e0e8f4bb281e", "63485fb688def0de08ae284fbc0cb6b7d9a6696e9542eb9328d69de33607f902", "6767b23b636e9b89c8b3ac3614fbdf529b6485cb3fc8ae110b705dbcb9e1fe12", "b54c6a95a349be5a1c09e487b4149111c6ccfbdaec5ec6f8ff82fb21ee894113", "604b619329ec489482dcc4af6f5a02f9ed1418f624f900eca143355127406217", "a556c534da5f5549bceb1ba95535dd57af5fd59011772552d5fea78a02e78a11", "34eae6950d3227d848e7e075aecdb65791306156b6503a89406e5802bfd0e117", "a6f3a081dd622f89f0e8703281052ff2a275507214fed44d3813a5027e5c762a"]
```
Comment (jm-clius) [inline tests/v2/test_waku_rln_relay.nim:303]: Wouldn't it be possible to generate this `seq` in a test utility function? I imagine it should even be possible to write a repeatable generator with the same result each time.
Otherwise, perhaps the `seq` can be extracted as a `const` in a separate module and reused between tests, because I see it's initialized elsewhere as well?

File: tests/v2/test_waku_rln_relay.nim [hunk 2]
```diff
@@ -666,4 +723,29 @@ suite "Waku rln relay":
...
```
Comment (jm-clius) [inline tests/v2/test_waku_rln_relay.nim:738]: See comment above. Will be a bit more readable if this is constructed elsewhere, if possible. :)

File: waku/v2/node/wakunode2.nim [hunk 3]
```diff
@@ -573,11 +584,40 @@ proc mountRelay*(node: WakuNode,
...
+      # validate the user-supplied membership index
+      if rlnRelayMemIndex < uint(0) or rlnRelayMemIndex >= uint(groupSize):
+        error "wrong membership index, failed to mount WakuRLNRelay"
+      else: 
+        # prepare group related inputs from the hardcoded keys
+        var groupKeyPairs = groupKeys.toMembershipKeyPairs()
+        var groupIDCommitments = groupKeyPairs.mapIt(it.idCommitment)
+
+        # mount rlnrelay in offline mode
+        waitFor node.mountRlnRelay(groupOpt= some(groupIDCommitments), memKeyPairOpt = some(groupKeyPairs[rlnRelayMemIndex]), memIndexOpt= some(rlnRelayMemIndex), onchainMode = false)
```
Comment (jm-clius) [inline waku/v2/node/wakunode2.nim:605]: Could these arguments be constructed in a util function, perhaps outside the `wakunode2` module (e.g. in `waku_setup`, or similar)? I'm worried that the `wakunode2` module may become unwieldy, especially since this is the entry point when debugging issues.

File: tests/v2/test_waku_rln_relay.nim (hunk 1 above)
Comment (staheri14) [reply tests/v2/test_waku_rln_relay.nim]: The underlying key generation is non-deterministic and each time produces a different result, I also do not have access to the randomness used in that proc to be able to control the output. But I can certainly use the `const` idea, Thanks! done in this [commit](https://github.com/status-im/nim-waku/pull/718/commits/24126e875185422d16e502f22528d319ae541cb4)

File: tests/v2/test_waku_rln_relay.nim (hunk 2 above)
Comment (staheri14) [reply tests/v2/test_waku_rln_relay.nim:738]: Thanks, done in https://github.com/status-im/nim-waku/pull/718/commits/24126e875185422d16e502f22528d319ae541cb4! :)

File: waku/v2/node/wakunode2.nim (hunk 3 above)
Comment (staheri14) [reply waku/v2/node/wakunode2.nim:605]: Can you please elaborate on the use of a util func to construct arguments?

Comment (jm-clius) [reply waku/v2/node/wakunode2.nim:605]: I meant something like the [`setupNat` function](https://github.com/status-im/nim-waku/blob/dcba72677e472e93b03c17d22e1ca7d427d953c7/waku/v2/node/wakunode2.nim#L856) where the arguments (e.g. groupOpt, memKeyPairOpt) are constructed in a util function in a separate module to keep `wakunode2` relatively uncluttered. However, now that `StaticGroupKeys` live elsewhere it's already much clearer and probably not necessary. Thanks! :)

File: waku/v2/node/wakunode2.nim [hunk 4]
```diff
@@ -573,11 +584,40 @@ proc mountRelay*(node: WakuNode,
...
     if rlnRelayEnabled:
-      # TODO pass rln relay inputs to this proc, right now it uses default values that are set in the mountRlnRelay proc
+      # TODO get user inputs via cli options
       info "WakuRLNRelay is enabled"
-      waitFor mountRlnRelay(node)
-      info "WakuRLNRelay is mounted successfully"
 
+      # a static list of 50 membership keys in hexadecimal format
+      var groupKeys = StaticGroupKeys
+      var groupSize = int(groupKeys.len/2)
```
Comment (jm-clius) [inline waku/v2/node/wakunode2.nim:592]: For interest's sake, any specific reason these are `var` and not `let`? I may be missing it, but I don't think these variables are modified after construction in which case `let` will be more appropriate. See [style guide](https://status-im.github.io/nim-style-guide/language.vardecl.html#variable-declaration-languagevardecl)

File: tests/v2/test_waku_rln_relay.nim [hunk 5]
```diff
@@ -666,4 +723,29 @@ suite "Waku rln relay":
...
     # verification of the bad proof should fail
-    doAssert(badF == 1)
\ No newline at end of file
+    doAssert(badF == 1)
+  
+  test "create a list of membership keys and construct a Merkle tree based on the list":
+    let (list, root) = createMembershipList(100) 
+    debug "created membership key list", list
+    check: 
+      list.len == 200 # two parts for each membership key i.e., one id key and one id commitment
```
Comment (oskarth) [inline tests/v2/test_waku_rln_relay.nim:732]: If there are really 100 logical elements in this list, perhaps it should be stored as a tuple instead? So `[(key0, commitment0), (key1, commitment2), ...]`. Right now it isn't obvious to me how the 200 items are mapped

File: tests/v2/test_waku_rln_relay.nim [hunk 6]
```diff
@@ -666,4 +723,29 @@ suite "Waku rln relay":
...
+    debug "the Merkle tree root", root
+    check:
+      root.len == 64
```
Comment (oskarth) [inline tests/v2/test_waku_rln_relay.nim:735]: Why is this 64? Comment above explaining what this is ensuring

File: tests/v2/test_waku_rln_relay.nim [hunk 7]
```diff
@@ -666,4 +723,29 @@ suite "Waku rln relay":
...
+  
+  test "check correctness of toMembershipKeyPairs and calcMerkleRoot":
+    var groupKeys: seq[string] = StaticGroupKeys
+    var groupKeyPairs = groupKeys.toMembershipKeyPairs()
+    debug "groupKeyPairs", groupKeyPairs
+    check: groupKeyPairs.len == 50
```
Comment (oskarth) [inline tests/v2/test_waku_rln_relay.nim:741]: Ditto above

File: tests/v2/test_waku_rln_relay.nim [hunk 8]
```diff
@@ -666,4 +723,29 @@ suite "Waku rln relay":
...
+
+    var groupIDCommitments = groupKeyPairs.mapIt(it.idCommitment)
+    debug "groupIDCommitments", groupIDCommitments
+    check: 
+      groupIDCommitments.len == 50
+
+    var expectedRoot = calcMerkleRoot(groupIDCommitments)
+    debug "expectedRoot", expectedRoot
+
+    check: expectedRoot == "65b753df62fb9a40575b116ba4138a864c66267357fdfca11db82de8bd73b400"
```
Comment (oskarth) [inline tests/v2/test_waku_rln_relay.nim:751]: What is this?

File: waku/v2/node/config.nim [hunk 9]
```diff
@@ -75,6 +75,11 @@ type
       defaultValue: false
       name: "rln-relay" }: bool
     
+    rlnRelayMemIndex* {.
+      desc: "the index of node in the rln-relay group: a value between 0-49 inclusive",
```
Comment (oskarth) [inline waku/v2/node/config.nim:79]: Presumably this is temporary, so might want to indicate this with something like `(for development)` `(experimental)` or so

File: waku/v2/node/wakunode2.nim [hunk 10]
```diff
@@ -573,11 +584,40 @@ proc mountRelay*(node: WakuNode,
...
+
+        info "membership id key", idkey=groupKeyPairs[rlnRelayMemIndex].idKey.toHex
+        info "membership id commitment key", idCommitmentkey=groupIDCommitments[rlnRelayMemIndex].toHex
+
+        # check the correct construction of the tree by comparing the calculated root against the expected root
+        # no error should happen as it is already captured in the unit tests
+        # TODO have added this check to account for unseen corner cases, will remove it later 
+        let 
+          root = node.wakuRlnRelay.rlnInstance.getMerkleRoot.value.toHex() 
+          expectedRoot = "65b753df62fb9a40575b116ba4138a864c66267357fdfca11db82de8bd73b400"
```
Comment (oskarth) [inline waku/v2/node/wakunode2.nim:615]: Where is this from? Comment

File: waku/v2/protocol/waku_rln_relay/waku_rln_relay_types.nim [hunk 11]
```diff
@@ -42,6 +42,8 @@ type WakuRLNRelay* = object
 # TODO may be able to make these constants private and put them inside the waku_rln_relay_utils
 const 
   MembershipFee* = 5.u256
-  Depth* = 32.u256
+  #  the current implementation of the rln lib only supports a circuit for Merkle tree with depth 32
+  MerkleTreeDepth* = 32
```
Comment (oskarth) [inline waku/v2/protocol/waku_rln_relay/waku_rln_relay_types.nim:46]: I think we can probably set this to 20

Also I saw 64 before but not sure what that is referring to?

File: waku/v2/protocol/waku_rln_relay/waku_rln_relay_types.nim [hunk 12]
```diff
@@ -42,6 +42,8 @@ type WakuRLNRelay* = object
...
...1a8ae18bad00c35597010a2107cad5041ffab69329", "d73e8c0e36e5588836bc9be4820e108d008f8399e4a9bacb5a452d4850a72d10", "01b81a331bc0c7818812c3848d382551183b35a16d74df5c8c32d561edb74b1c", "403e3cb57be08aef37dcc8af0a93a6ea8064540c8bcdb76fb34875c3d894be03", "6e420bc668f7b2aeee7e719daa3a7a4471bd04e8bcf24a27e4a5ae07e8c44207", "98baf438da4c655b9d92dc9033ae0f5cd7f0efda9b0d1bb894d08f1ce8620a01", "1282dda3446fb157e236155197def0d309eecb9dde8445f599ead69d4f31d916", "6453a3cf73ab5f3bafe645ce69fcdb612526fa66659e564c783c5600fb3a400f", "a8c1bfdbae2f56d10ccaf89887286c7c3256f7bdba21e98ac3c44578926c642c", "34bd74a58df94ae160f90bb64d202908d95b9dd0d89f02cd7e6bc8ddaf11280f", "a7377899f99a93796cd8901e17357403a900570d869fc49afed5e0e8f4bb281e", "63485fb688def0de08ae284fbc0cb6b7d9a6696e9542eb9328d69de33607f902", "6767b23b636e9b89c8b3ac3614fbdf529b6485cb3fc8ae110b705dbcb9e1fe12", "b54c6a95a349be5a1c09e487b4149111c6ccfbdaec5ec6f8ff82fb21ee894113", "604b619329ec489482dcc4af6f5a02f9ed1418f624f900eca143355127406217", "a556c534da5f5549bceb1ba95535dd57af5fd59011772552d5fea78a02e78a11", "34eae6950d3227d848e7e075aecdb65791306156b6503a89406e5802bfd0e117", "a6f3a081dd622f89f0e8703281052ff2a275507214fed44d3813a5027e5c762a"]
```
Comment (oskarth) [inline waku/v2/protocol/waku_rln_relay/waku_rln_relay_types.nim:49]: Comment explaining this is temporary and why it is like this (e.g. link to issue to make it dynamic, or just saying this is temp for testing and was generated by <what procedure>)

File: waku/v2/protocol/waku_rln_relay/waku_rln_relay_utils.nim [hunk 13]
```diff
@@ -130,3 +130,34 @@ proc getMerkleRoot*(rlnInstance: RLN[Bn256]): MerkleNodeResult =
   var rootValue = cast[ptr array[32,byte]] (root.`ptr`)
   let merkleNode = rootValue[]
   return ok(merkleNode)
+
+proc toMembershipKeyPairs*(groupKeys: seq[string]): seq[MembershipKeyPair] {.raises: [Defect, ValueError]} =
+  ## groupKeys is an alternating sequence of identity keys and their corresponding id commitment keys in the hexadecimal format
```
Comment (oskarth) [inline waku/v2/protocol/waku_rln_relay/waku_rln_relay_utils.nim:135]: Could this be a list of tuples instead? This seems like a pretty ad hoc data structure that will be prone to bugs

**Review by oskarth — COMMENTED** [review]
Thanks, nice to see this getting close!

Left comments in-line. In general, the biggest problem is that there are a lot of unexplained magic constants in the code. That they are needed makes sense, especially at a PoC stage, but they can probably be encoded in a better way. The goal should be that anyone can read code and understand where things come from. Specifically:

1) Turn magic numbers into symbolic constants
2) Add comments for why they are there and where they come from

So instead of writing, as an example say 32 and 64, define 32 as `MERKLE_DEPTH` and then use `MERKLE_DEPTH * 2` (with comment). Etc.

Here's a specific example of how to do it from Nimbus: https://github.com/status-im/nimbus-eth2/blob/stable/beacon_chain/spec/presets/v1_0_1/mainnet.nim

**Comment by D4nte** [comment]
Hi @staheri14, sorry but I do not have enough knowledge about RLN to add any value to this review at this time.
Is there anything specific you want me to look at?

File: waku/v2/node/wakunode2.nim (hunk 4 above)
Comment (staheri14) [reply waku/v2/node/wakunode2.nim:592]: Right, nice catch! forgot to take them back to `let`  :),  I defined them as `var` since I needed to update their content afterward, later I have changed the logic but forgot to edit this part. Thanks for the comment!

Comment (staheri14) [reply waku/v2/node/wakunode2.nim]: done in https://github.com/status-im/nim-waku/commit/925f5863488363761526a874dfd54c9aac65d275

**Comment by oskarth** [comment]
@staheri14 let me know once above comments have been addressed and you want a review

File: waku/v2/node/wakunode2.nim (hunk 3 above)
Comment (staheri14) [reply waku/v2/node/wakunode2.nim:607]: Thanks for the explanation! 
I'll certainly use a util function in case that the input preparation gets more complicated (in the future versions), thanks!

File: waku/v2/protocol/waku_rln_relay/waku_rln_relay_utils.nim (hunk 13 above)
Comment (staheri14) [reply waku/v2/protocol/waku_rln_relay/waku_rln_relay_utils.nim]: Done!

File: waku/v2/protocol/waku_rln_relay/waku_rln_relay_types.nim (hunk 12 above)
Comment (staheri14) [reply waku/v2/protocol/waku_rln_relay/waku_rln_relay_types.nim]: Sure, done!

File: waku/v2/protocol/waku_rln_relay/waku_rln_relay_types.nim (hunk 11 above)
Comment (staheri14) [reply waku/v2/protocol/waku_rln_relay/waku_rln_relay_types.nim]: As for the tree height, I am still communicating with the rln team, once I figure out whether the size is fixed or not, I will update this part as well.

64 was the size of the Poseidon hash output in hex, in my new commits I replaced it with a proper const variable.

File: waku/v2/node/wakunode2.nim (hunk 10 above)
Comment (staheri14) [reply waku/v2/node/wakunode2.nim]: sure, done!

File: waku/v2/node/config.nim (hunk 9 above)
Comment (staheri14) [reply waku/v2/node/config.nim]: Thanks, done!

File: tests/v2/test_waku_rln_relay.nim (hunk 6 above)
Comment (staheri14) [reply tests/v2/test_waku_rln_relay.nim]: Sure, done!

File: tests/v2/test_waku_rln_relay.nim (hunk 5 above)
Comment (staheri14) [reply tests/v2/test_waku_rln_relay.nim]: Revised it to a list of 100 tuples, the new version should be more clear. Please let me know if there is unclarity.

File: tests/v2/test_waku_rln_relay.nim [hunk 14]
```diff
@@ -259,6 +285,37 @@ procSuite "Waku rln relay":
...
+    debug "groupIDCommitments", groupIDCommitments
+    var expectedRoot = calcMerkleRoot(groupIDCommitments)
+    debug "expectedRoot", expectedRoot
+   
+    # start rln-relay in off-chain mode
+    await node.mountRlnRelay(groupOpt = some(groupIDCommitments), memKeyPairOpt = some(groupKeyPairs[index]),  memIndexOpt = some(index), onchainMode = false)
+    
+    let calculatedRoot = node.wakuRlnRelay.rlnInstance.getMerkleRoot().value().toHex
+    debug "calculated root by wakuRlnRelay", calculatedRoot
+    check expectedRoot == calculatedRoot
```
[the remark this answers is not in the document]
Comment (staheri14) [reply tests/v2/test_waku_rln_relay.nim]: Of course, done!

File: tests/v2/test_waku_rln_relay.nim [hunk 15]
```diff
@@ -259,6 +285,37 @@ procSuite "Waku rln relay":
...
+    await node.start()
+
+    var rlnInstance = createRLNInstance()
+    check rlnInstance.isOk == true
+    var rln = rlnInstance.value
+
+    # current peer index in the Merkle tree
+    let index = uint(5)
+
+    # Create a group of 50 members
```
[the remark this answers is not in the document]
Comment (staheri14) [reply tests/v2/test_waku_rln_relay.nim]: Clarified in the latest commits.

File: tests/v2/test_waku_rln_relay.nim [hunk 16]
```diff
@@ -259,6 +285,37 @@ procSuite "Waku rln relay":
...
+      node = WakuNode.new(nodeKey, ValidIpAddress.init("0.0.0.0"),
+        Port(60000))
+    await node.start()
+
+    var rlnInstance = createRLNInstance()
+    check rlnInstance.isOk == true
+    var rln = rlnInstance.value
+
+    # current peer index in the Merkle tree
+    let index = uint(5)
```
[the remark this answers is not in the document]
Comment (staheri14) [reply tests/v2/test_waku_rln_relay.nim]: Sure, fixed by adding more comments. Please let me know if the new comments help.

**Comment by staheri14** [comment]
> The goal should be that anyone can read code and understand where things come from. 

Fair enough! :)

> Specifically:
>  1. Turn magic numbers into symbolic constants
> 2. Add comments for why they are there and where they come from

Thanks @oskarth for your constructive comments! 
I revised the code and defined constant variables for the recurring values (which meant something in the context of waku-rln-relay) and added clarifying comments everywhere I found ambiguity (including the parts you highlighted in your comments).
Please let me know if I have missed something.

**Comment by staheri14** [comment]
> Hi @staheri14, sorry but I do not have enough knowledge about RLN to add any value to this review at this time.
> Is there anything specific you want me to look at?

@D4nte 
As rln-relay is getting closer to the dogfooding stage, I thought you might be interested to be informed about its latest development and how its CLI looks like :) feel free to comment on any aspect of it.

**Review by oskarth — APPROVED** [review]

**Review by jm-clius — APPROVED** [review]
Good to see this coming together. 👍
